src/key.js

```diff
@@ -0,0 +1,314 @@
+/* @flow */
+'use strict'
+
+const path = require('path')
+const uuid = require('uuid/v4')
+
+/**
+ * A Key represents the unique identifier of an object.
+ * Our Key scheme is inspired by file systems and Google App Engine key model.
+ * Keys are meant to be unique across a system. Keys are hierarchical,
+ * incorporating more and more specific namespaces. Thus keys can be deemed
+ * 'children' or 'ancestors' of other keys:
+ * - `new Key('/Comedy')`
+ * - `new Key('/Comedy/MontyPython')`
+ * Also, every namespace can be parametrized to embed relevant object
+ * information. For example, the Key `name` (most specific namespace) could
+ * include the object type:
+ * - `new Key('/Comedy/MontyPython/Actor:JohnCleese')`
+ * - `new Key('/Comedy/MontyPython/Sketch:CheeseShop')`
+ * - `new Key('/Comedy/MontyPython/Sketch:CheeseShop/Character:Mousebender')`
+ *
+ */
```

Comment:
this is jsdoc, will go into the generated api docs, as we do with other api docs